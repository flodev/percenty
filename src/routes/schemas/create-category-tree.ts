import { catchError, forkJoin, map, mergeMap, of, tap, throwError, type Observable } from 'rxjs';
import { Category } from './category.svelte';
import type { CategorySchemaType } from './schema';

export function createCategoryTree(
  data: CategorySchemaType,
  parent?: Category
): Observable<Category> {
  return of(data).pipe(
    // Transform the input data into a Category object
    map((currentData) => {
      return makeClass(currentData, parent);
    }),

    // Process subcategories recursively
    mergeMap((cat) => {
      if (cat.categories.length > 0) {
        // Recursively process all subcategories in parallel
        return forkJoin(cat.categories.map((newCat) => createCategoryTree(newCat, cat))).pipe(
          // Return the parent category with processed subcategories
          map((processedSubcategories) => {
            cat.categories = processedSubcategories;
            return cat;
          })
        );
      } else {
        // No subcategories, return the category as-is
        return of(cat);
      }
    }),

    // Handle errors
    catchError((e) => {
      console.error('failed:', e);
      return throwError(() => new Error('Processing failed'));
    })
  );
}

function makeClass(category: CategorySchemaType, parent?: Category) {
  return new Category(
    category.name,
    category.percent,
    category.categories,
    category.percentages,
    parent,
    parent == undefined ? category.amount : undefined
  );
}
