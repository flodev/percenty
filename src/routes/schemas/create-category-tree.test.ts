import { describe, expect, it } from 'vitest';
import { createCategoryTree } from './create-category-tree';
import type { CategorySchemaType } from './schema';
import { lastValueFrom } from 'rxjs';

describe('create category tree', () => {
  const testData: CategorySchemaType = {
    amount: 1000,
    name: 'root',
    categories: [
      {
        name: 'risk part',
        percent: 80,
        categories: [
          {
            name: 'ETFs',
            percent: 70,
            categories: [],
            percentages: [
              { name: 'world', percent: 70 },
              { name: 'EM', percent: 30 }
            ]
          },
          {
            name: 'sustainable',
            percent: 20,
            categories: [],
            percentages: [{ name: 'ETF sustainable', percent: 100 }]
          },
          {
            name: 'commondities',
            percent: 10,
            categories: [],
            percentages: [{ name: 'ETF commodities', percent: 100 }]
          }
        ],
        percentages: []
      },
      {
        name: 'low risk',
        percent: 20,
        categories: [],
        percentages: [{ name: 'government bond', percent: 100 }]
      }
    ],
    percentages: [],
    percent: 100
  };
  it('creates the tree', async () => {
    const lastValue = await lastValueFrom(createCategoryTree(testData));
    expect(lastValue).toMatchSnapshot();
  });
  it('the leafes will still calculate the amount properly', async () => {
    const lastValue = await lastValueFrom(createCategoryTree(testData));
    expect(lastValue.categories[0].categories[0].amount).toEqual(560);
  });
});
