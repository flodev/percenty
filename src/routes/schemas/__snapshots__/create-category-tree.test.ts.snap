// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create category tree > creates the tree 1`] = `
{
  "amount": 1000,
  "categories": [
    {
      "amount": undefined,
      "categories": [
        {
          "amount": undefined,
          "categories": [],
          "name": "ETFs",
          "percent": 70,
          "percentages": [
            {
              "name": "world",
              "percent": 70,
            },
            {
              "name": "EM",
              "percent": 30,
            },
          ],
        },
        {
          "amount": undefined,
          "categories": [],
          "name": "sustainable",
          "percent": 20,
          "percentages": [
            {
              "name": "ETF sustainable",
              "percent": 100,
            },
          ],
        },
        {
          "amount": undefined,
          "categories": [],
          "name": "commondities",
          "percent": 10,
          "percentages": [
            {
              "name": "ETF commodities",
              "percent": 100,
            },
          ],
        },
      ],
      "name": "risk part",
      "percent": 80,
      "percentages": [],
    },
    {
      "amount": undefined,
      "categories": [],
      "name": "low risk",
      "percent": 20,
      "percentages": [
        {
          "name": "government bond",
          "percent": 100,
        },
      ],
    },
  ],
  "name": "root",
  "percent": 100,
  "percentages": [],
}
`;
